azerothcore/azerothcore-wotlk
# feat(Core/Modules): add separated lib for modules

## Changes Proposed:
1. General
- Make sepated library `modules`
- Make new manager `ModuleMgr` for check enable modules and other
- Enable loading modules configs if module enable. (Previously, all configurations were always loaded regardless of whether the module was disabled or not)
2. Eluna
- Move all `ELUNA` code to module
- Added new hooks for support `mod-eluna`
- Fix using double core hooks

## Issues Addressed:
- Part dynamic linking

## Tests Performed:
- In game

## How to Test the Changes:
1. Startup core with/without modules
2. Check any scripts
3. Check any modules
4. Check eluna scripts with https://github.com/azerothcore/mod-eluna-lua-engine/pull/71

## Linked PR's:
- [ ] https://github.com/azerothcore/mod-eluna-lua-engine/pull/71

## Media:
- Add new info in command `.ser debug`
![image](https://user-images.githubusercontent.com/18329778/142872837-fd912ad7-3023-42b2-89d3-2d2b43a5e1de.png)

- New structure for modules
![image](https://user-images.githubusercontent.com/18329778/142872961-3061764e-b6e7-41dc-ab33-6df8bc8ec384.png)

## How to Test AzerothCore PRs
 
When a PR is ready to be tested, it will be marked as **[WAITING TO BE TESTED]**.

You can help by testing PRs and writing your feedback here on the PR's page on GitHub. Follow the instructions here:

http://www.azerothcore.org/wiki/How-to-test-a-PR

**REMEMBER**: when testing a PR that changes something **generic** (i.e. a part of code that handles more than one specific thing), the tester should not only check that the PR does its job (e.g. fixing spell XXX) but **especially** check that the PR does not cause any regression (i.e. introducing new bugs).

**For example**: if a PR fixes spell X by changing a part of code that handles spells X, Y, and Z, we should not only test X, but **we should test Y and Z as well**.
